llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -2553,6 +2553,47 @@ void CombinerHelper::applyCombineTruncOfExt(
   MI.eraseFromParent();
 }
 
+bool CombinerHelper::matchCombineShlOfAnd(MachineInstr &MI, Register &Reg) {
+  // We're trying to match the following pattern:
+  //   %t = G_AND %x, (2^imm - 1)
+  //   %root = G_SHL %t, imm
+  // -->
+  //   %root = G_SHL %x, imm
+  // Where imm >= half size of %x
+  assert(MI.getOpcode() == TargetOpcode::G_SHL && "Expected a G_SHL");
+
+  const Register SrcReg = MI.getOperand(1).getReg();
+  const Register ShftReg = MI.getOperand(2).getReg();
+  const LLT SrcTy = MRI.getType(SrcReg);
+  const unsigned Size = SrcTy.getSizeInBits();
+
+  const auto Shift = getIConstantVRegSExtVal(ShftReg, MRI);
+  if (!Shift)
+    return false;
+  const unsigned ShiftVal = Shift.value();
+  if (ShiftVal < Size / 2)
```

Comment:
I think this can be more general. Shift discards 'imm' upper bits of its source value, so if the AND only touches those bits, it is redundant. That means that the AND can be removed if its immediate has only one bits in the lower wordsize - imm bits. effectively, `(~andimm) << shiftimm == 0`.

Reply:
And this generalizes to shiftright (both versions) with very little change.

Reply:
Modified shiftleft accordingly, but there a combine for shiftright which does similar.
`matchBitfieldExtractFromShrAnd`